File: packages/sdk-flutter/lib/src/cache.dart
import 'config.dart';

class CacheResult {
  final Map<String, bool>? flags;
  final bool stale;
  final bool found;

  CacheResult({this.flags, this.stale = false, this.found = false});
}

class CacheStats {
  int hits;
  int misses;
  int staleHits;

  CacheStats({this.hits = 0, this.misses = 0, this.staleHits = 0});
}

class FlagCache {
  final CacheConfig _config;
  Map<String, bool>? _flags;
  DateTime? _timestamp;
  final CacheStats _stats = CacheStats();

  FlagCache(this._config);

  CacheResult get() {
    if (_flags == null || _timestamp == null) {
      _stats.misses++;
      return CacheResult(found: false);
    }

    final age = DateTime.now().difference(_timestamp!);

    if (age > _config.staleTtl) {
      _stats.misses++;
      _flags = null;
      _timestamp = null;
      return CacheResult(found: false);
    }

    if (age > _config.ttl) {
      _stats.staleHits++;
      return CacheResult(
        flags: Map<String, bool>.from(_flags!),
        stale: true,
        found: true,
      );
    }

    _stats.hits++;
    return CacheResult(
      flags: Map<String, bool>.from(_flags!),
      stale: false,
      found: true,
    );
  }

  void set(Map<String, bool> flags) {
    _flags = Map<String, bool>.from(flags);
    _timestamp = DateTime.now();
  }

  void clear() {
    _flags = null;
    _timestamp = null;
  }

  bool hasAny() {
    if (_flags == null || _timestamp == null) return false;
    return DateTime.now().difference(_timestamp!) <= _config.staleTtl;
  }

  CacheStats getStats() => CacheStats(
        hits: _stats.hits,
        misses: _stats.misses,
        staleHits: _stats.staleHits,
      );

  void recordHit(bool stale) {
    if (stale) {
      _stats.staleHits++;
    } else {
      _stats.hits++;
    }
  }

  void recordMiss() {
    _stats.misses++;
  }
}

File: packages/sdk-dotnet/src/Rollgate.SDK/Cache.cs
namespace Rollgate.SDK;

public class CacheResult
{
    public Dictionary<string, bool>? Flags { get; set; }
    public bool Stale { get; set; }
    public bool Found { get; set; }
}

public class CacheStats
{
    public long Hits { get; set; }
    public long Misses { get; set; }
    public long StaleHits { get; set; }
}

public class FlagCache
{
    private readonly CacheConfig _config;
    private readonly object _lock = new();
    private Dictionary<string, bool>? _flags;
    private DateTime _timestamp;
    private long _hits;
    private long _misses;
    private long _staleHits;

    public FlagCache(CacheConfig config)
    {
        _config = config;
    }

    public CacheResult Get()
    {
        lock (_lock)
        {
            if (_flags == null)
            {
                _misses++;
                return new CacheResult { Found = false };
            }

            var age = DateTime.UtcNow - _timestamp;

            if (age > _config.StaleTtl)
            {
                _misses++;
                _flags = null;
                return new CacheResult { Found = false };
            }

            if (age > _config.Ttl)
            {
                _staleHits++;
                return new CacheResult
                {
                    Flags = new Dictionary<string, bool>(_flags),
                    Stale = true,
                    Found = true
                };
            }

            _hits++;
            return new CacheResult
            {
                Flags = new Dictionary<string, bool>(_flags),
                Stale = false,
                Found = true
            };
        }
    }

    public void Set(Dictionary<string, bool> flags)
    {
        lock (_lock)
        {
            _flags = new Dictionary<string, bool>(flags);
            _timestamp = DateTime.UtcNow;
        }
    }

    public void Clear()
    {
        lock (_lock)
        {
            _flags = null;
        }
    }

    public bool HasAny()
    {
        lock (_lock)
        {
            if (_flags == null) return false;
            return (DateTime.UtcNow - _timestamp) <= _config.StaleTtl;
        }
    }

    public CacheStats GetStats()
    {
        lock (_lock)
        {
            return new CacheStats
            {
                Hits = _hits,
                Misses = _misses,
                StaleHits = _staleHits
            };
        }
    }

    public void RecordHit(bool stale)
    {
        lock (_lock)
        {
            if (stale) _staleHits++;
            else _hits++;
        }
    }

    public void RecordMiss()
    {
        lock (_lock)
        {
            _misses++;
        }
    }
}
